jenkins/automation/functions

```diff
@@ -0,0 +1,276 @@
+  local remote_name="${1?remote name not specified}"
+  local base_branch="${2?base branch not specified}"
+  local image_fullname="${3?image not specified}"
+
+  local image_name="${image_fullname%:*}"
+  local image_tag="${image_fullname##*:}"
+  local image_revision=${image_tag##*-r}
+  local app_name=${image_name##*/}
+  local app_version="${image_tag%%-*}"
+  local app_version_series="${app_version%.*}"
+  local dev_branch_prefix="autoupdate-${base_branch}-${app_name}"
+  local dev_branch="${dev_branch_prefix}-${app_version}"
+
+  pr_exists=0
+  if branch_exists "${remote_name}/${dev_branch}"; then
+    # if branch exists, it implies that a PR already exists
+    pr_exists=1
+
+    # checkout to the existing branch so that branch checks are proper in the next step
+    checkout_branch "${remote_name}" "${dev_branch}"
+  fi
+
+  local branch_image_tag="$(sed -n -e "s|.*\"${image_name}:\([0-9]*.[0-9]*.[0-9]*-r[0-9]*\)\".*|\1|p" "${MANIFESTS_IMAGES_JSON}")"
+  local branch_image_revision=${branch_image_tag##*-r}
+  local branch_app_version="${branch_image_tag%%-*}"
+  local branch_app_version_series="${branch_app_version%.*}"
+
+  if [[ -z "${branch_image_tag}" ]]; then
+    # image is not used, nothing to do!
+    return 0
+  fi
+
+  if [[ $(vercmp "${app_version}" "${branch_app_version}") -lt 0 ]]; then
+    # current version is higher then specified, nothing to do!
+    return 0
+  fi
+
+  if [[ $(vercmp "${app_version}" "${branch_app_version}") -eq 0 ]]; then
+    # app version is the same, can only update existing pr's
+    if [[ $(vercmp "${image_revision}" "${branch_image_revision}") -le 0 ]]; then
+      # image revision is the same, nothing to do!
+      return 0
+    elif [[ "${pr_exists}" -eq 0 ]]; then
+      # image revision has updated, but skip_revision_updates is set
+      return 0
+    fi
+  fi
+
+  # release branches only permit PATCH version updates
+  if [[ "${base_branch}" =~ ${RELEASES_BRANCH_GLOB} ]] && [[ $(vercmp "${app_version_series}" "${branch_app_version_series}") -ne 0 ]]; then
+    return 1
+  fi
+
+  info "[${base_branch}] New image '${image_fullname}' available"
+
+  if [[ "${pr_exists}" -eq 0 ]]; then
+    info "[${base_branch}] Creating branch '${dev_branch}'"
+    create_branch "${base_branch}" "${dev_branch}"
+  else
+    info "[${base_branch}] Will update existing '${dev_branch}' branch..."
+  fi
+
+  # apply the update and commit the changes
+  sed -i "s|${image_name}:[0-9]*\.[0-9]*\.[0-9]*-r[0-9]*|${image_fullname}|" "${MANIFESTS_IMAGES_JSON}"
+  commit_changes "${MANIFESTS_IMAGES_JSON}" "${app_name}: component image updated to '${image_fullname}'"
+
+  info "[${base_branch}] Pushing changes to '${dev_branch}'"
+  push_changes "${remote_name}" "${dev_branch}"
+
+  if [[ "${pr_exists}" -eq 0 ]]; then
+    info "[${base_branch}] Creating pull request with '${dev_branch}'"
+    create_pull_request "${base_branch}" "[maintenance/${base_branch}] '${app_name}' updated to '${app_version}'"
```

Comment:
Looking at the generated PRs, I found myself wanting to know the "severity" of the upgrade, before digging through to the actual diff:

```suggestion
    create_pull_request "${base_branch}" "[maintenance/${base_branch}] '${app_name}' updated '${branch_app_version}' -> '${app_version}'"
```

Reply:
fair enough!